### ISSUE TITLE
Problem with ThresholdFilter and GaussianScorer

### ISSUE THREAD

**Issue by dberardo-com** [online-ml/river#948]
i have got this signal: 

![image](https://user-images.githubusercontent.com/65530457/172017334-35c2d50a-9200-4845-8f44-64c2dea33cc8.png)

and i am applying a gaussianScorer  + ThresholdFilter like in the timeseries Waterflow example.

The problem is that at the end of the execution, the gaussian filter has not learnt anything that happened after the jump at the beginning of April. Here is how the plot of the score_one(x,y) function looks like:

![image](https://user-images.githubusercontent.com/65530457/172017418-0021458d-2a63-4377-9c02-d705227d5a9d.png)

1 means anomaly ... so you can see that at the end of the execution, the gaussian scorer has not recognized the dirft ... although it had quite a few days to do so:

![image](https://user-images.githubusercontent.com/65530457/172017514-1e25e119-4969-48ce-a61b-31810ba954dc.png)

here you can see that the mean is stuck at 161, which is close to the initial value for the first couple of weeks

**Comment by dberardo-com** [comment]
as an extra information, if i change the parameter window_size to 48, this is what i get:

![image](https://user-images.githubusercontent.com/65530457/172017883-5be94d34-8382-423f-b6a9-a40ea7006f18.png)

![image](https://user-images.githubusercontent.com/65530457/172017908-f61c4032-660c-4698-9c02-921445d9270e.png)


which means a slightly improved situation, but the adpatation is too slow here.


So my question here is: is the thresholder updating correctly or why is that change so slow?

**Comment by dberardo-com** [comment]
and here is what happens for a smaller window_size of 30:

![image](https://user-images.githubusercontent.com/65530457/172018017-0b3babb3-7c9d-43a1-81a5-ed31369d9c8b.png)

![image](https://user-images.githubusercontent.com/65530457/172018039-314d091d-83cd-46fc-8195-6854bdb1a1b9.png)


which makes no sense, as i would expect an reduced number of anomalies detected

--> indeed if i set the window_size to 1 i get always 0 as anomaly score, which is correct

actually i always get 0 score for values of window_size which are below 30 ... is there a strange relation between grace_period and window_size?

**Comment by dberardo-com** [comment]
alright, so to answer some of my own questions:

> actually i always get 0 score for values of window_size which are below 30 ... is there a strange relation between grace_period and window_size?

i have got answered here that grace_period cannot be larger the window_size: https://github.com/online-ml/river/blob/main/river/anomaly/gaussian.py#L71  i think i will try to patch this one because it makes sense sometimes to have a higher grace_period then window_size

> So my question here is: is the thresholder updating correctly or why is that change so slow?

i have just realized now that the parameter `protect_anomaly_detector` is set to True by default, which means that the GaussianScorer will never catch up the drift, so the above behavior is actually well justified. The problem is now that when i set that flag to False, the GaussianScorer actually never learns anything. This is the state of the model after training with the same dataset:

![image](https://user-images.githubusercontent.com/65530457/172052992-3dcd486c-e039-4795-bf42-c725480670a9.png)

Also, no matter the value of window size now, i always get a 0 when scoring anomalies this is clearly an unwanted behavior. is this happening to you as well in the waterFlow dataset example ? @MaxHalford

**Comment by dberardo-com** [comment]
ok, found the bug and created a PR: https://github.com/dberardo-com/river/pull/2

**Comment by dberardo-com** [comment]
after fixing this bug, and reducing the learning rate of the SNARIMAX regressor's optimizer i get a more useful prediction:

![image](https://user-images.githubusercontent.com/65530457/172053901-9e27f164-db7d-44bc-8450-dc77ae68da93.png)

still room for improvement though, i think i will now give it a try by using a GaussianScore with a proba.TimeRolling update function: https://github.com/online-ml/river/blob/main/river/proba/time_rolling.py

hopefully the detector will be even more robust then

**Comment by MaxHalford** [comment]
What's the status with the issue? Do you still want me to have a look at something @dberardo-com? :)

**Comment by dberardo-com** [comment]
no thanks, this is related to the open PR, so no more need for discussion here. probably a bit unrelated to this specific issue, but is there a place in the documentation where i can find out the best way to implement a "groupBy" hour aggregation for the input data points?

i mean something to go about this issue:

1. input data comes in at random intervals (every 10 second, sometimes with gaps, etc.) 
2. SNARIMAX wants hour timesteps
3. solution --> rolling mean with 1 hour window

but then the question is: when should the ARIMAX.learn_one method be called again? should it only happen really once every hour ? meaning maybe whenever the datetime().hour value changes, or is it no problem if the learn_one method is called each time a new data point comes in, because the rolling mean has not really changed that much?

i would like to include this in a test python script, so i can make a more realistic test-bed to include in the blog post

**Comment by MaxHalford** [comment]
> but then the question is: when should the ARIMAX.learn_one method be called again? should it only happen really once every hour ? meaning maybe whenever the datetime().hour value changes, or is it no problem if the learn_one method is called each time a new data point comes in, because the rolling mean has not really changed that much?

My gut feeling is that you have to accumulate the data, and update the model each time an hour is complete. I remember seeing someone have this issue before, but I can't remember where. Anyway the consensus was that you need an aggregation mechanism in front of your model.

> i would like to include this in a test python script, so i can make a more realistic test-bed to include in the blog post

Good idea!

**Comment by dberardo-com** [comment]
> > but then the question is: when should the ARIMAX.learn_one method be called again? should it only happen really once every hour ? meaning maybe whenever the datetime().hour value changes, or is it no problem if the learn_one method is called each time a new data point comes in, because the rolling mean has not really changed that much?
> 
> My gut feeling is that you have to accumulate the data, and update the model each time an hour is complete. I remember seeing someone have this issue before, but I can't remember where. Anyway the consensus was that you need an aggregation mechanism in front of your model.

mm i see, then i will have to just try it out. If you find any past conversation about the matter we can share this thoughts in a github Discussion. I guess i will be opening one when having the test results.

Just to finish this: is there any example how to use the TimeRolling Mean to get the average of the past hour in River? otherwise i will figure that out  by myself

**Comment by MaxHalford** [comment]
> mm i see, then i will have to just try it out. If you find any past conversation about the matter we can share this thoughts in a github Discussion. I guess i will be opening one when having the test results.

Added a card [here](https://www.notion.so/maxhalford/Find-a-solution-to-irregular-time-stamped-data-b9f4d90f633c4192babfee49c8dd61fe). I'll think about it, but it'll take time.

> Just to finish this: is there any example how to use the TimeRolling Mean to get the average of the past hour in River? otherwise i will figure that out by myself

See my comment [here](https://github.com/online-ml/river/pull/949#issuecomment-1156038373).

**Comment by dberardo-com** [comment]
https://github.com/online-ml/river/discussions/971

**Comment by dberardo-com** [comment]
a question on threshold filter here. how exactly is the pipeline supposed to understand that the filter has returned an anomaly and halt ?

i mean what is the signal to halt a pipeline from preventing the model underneath to learn from the incoming sample? i would expect some sort of CircuitBreaker exception to be thrown in code, but i cant find any in the thresholdfilter code.

what is the piece i am missing here?
